src/common/utils/registry/auth/authorizer.go

```diff
@@ -0,0 +1,122 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package auth
+
+import (
+	"fmt"
+	"github.com/docker/distribution/registry/client/auth/challenge"
+	"github.com/goharbor/harbor/src/common/http/modifier"
+	"net/http"
+	"net/url"
+	"strings"
+)
+
+// NewAuthorizer creates an authorizer that can handle different auth schemes
+func NewAuthorizer(credential Credential, client ...*http.Client) modifier.Modifier {
```

Comment:
1. Can we have a better name?
2. The attribute "url" is a little tricky, why not just set the registry endpoint when creating this authorizer?

Reply:
For 2, we handle the URL for a long time with this method and it works well. 
Besides this is the way how docker handles such case

Reply:
Could you make sure this authorizer is not for concurrent use?  Explicitly document it in the comment?

Reply:
Add lock to avoid the concurrent issue